File: minimum-insertion-steps-to-make-a-string-palindrome-1312/solution.js
/**
 * LeetCode problem # 1312, 'Minimum Insertion Steps to Make a String
 * Palindrome'
 * https://leetcode.com/problems/minimum-insertion-steps-to-make-a-string-palindrome
 */

/**
 * @param {string} s
 * @return {number}
 */
function minInsertions(s) {
  const n = s.length;

  const cache = Array(n + 1)
    .fill(null)
    .map((elem) => Array(n + 1).fill(-1));

  const solve = (i, j) => {
    if (j <= i + 1) {
      return 0;
    }
    if (cache[i][j] >= 0) {
      return cache[i][j];
    }

    if (s[i] === s[j - 1]) {
      cache[i][j] = solve(i + 1, j - 1);
      return cache[i][j];
    }
    const opt1 = 1 + solve(i, j - 1);
    const opt2 = 1 + solve(i + 1, j);
    cache[i][j] = Math.min(opt1, opt2);

    return cache[i][j];
  };

  return solve(0, n);
}

File: minimum-insertion-steps-to-make-a-string-palindrome-1312/Solution.java
/**
 * LeetCode problem # 1312, 'Minimum Insertion Steps to Make a String
 * Palindrome'
 * https://leetcode.com/problems/minimum-insertion-steps-to-make-a-string-palindrome
 */

class Solution {
    public int minInsertions(String s) {
        int lcs = longestPalindromeSubsequence(s);
        return s.length() - lcs;
    }

    public int longestPalindromeSubsequence(String s) {
        int n = s.length();
        int[][] dp = new int[n][n];
        for (int i = 0; i < n; ++i) {
            dp[i][i] = 1;
        }

        for (int i = n - 1; i >= 0; --i) {
            for (int j = i + 1; j < n; ++j) {
                if (s.charAt(i) == s.charAt(j)) {
                    dp[i][j] = dp[i + 1][j - 1] + 2;
                } else {
                    dp[i][j] = Math.max(dp[i + 1][j], dp[i][j - 1]);
                }
            }
        }
        return dp[0][n - 1];
    }
}
